[docs] Update common_issues.rst: update information about reveal type & locals (#19059)
python/mypy: docs/source/common_issues.rst

   #     a: builtins.int
   #     b: builtins.str
.. note::
   ``reveal_type`` and ``reveal_locals`` are understood by mypy during
   typechecking, and don't have to be imported as functions. However,
   if you don't import them, then they don't exist at runtime! Therefore,
   you'll have to remove any ``reveal_type`` and ``reveal_locals`` calls
   from your program or else Python will give you an error at runtime about
   those names being undefined. Alternatively, you can import ``reveal_type``

>    from ``typing_extensions`` (or, in more recent versions of python, from

Reviewer: `reveal_type` was added in 3.11, and we're now awaiting 3.14, so it isn't some shiny new thing. Maybe swap these parts? 

```
Alternatively, you can import ``reveal_type`` from ``typing`` (or, in versions of python older than 3.11, from ``typing_extensions``)
```

Author: I like it, thanks.

   #     b: builtins.str
.. note::

>     ``reveal_type`` and ``reveal_locals`` are understood by mypy during
>     typechecking, and don't have to be imported as functions. However, if you
>     don't import them, then they don't exist at runtime! Therefore, you'll have
>     to remove any ``reveal_type`` and ``reveal_locals`` calls from your program
>     or else Python will give you an error at runtime about those names being
>     undefined. Alternatively, you can import ``reveal_type`` from ``typing``
>     (or, in versions of python older than 3.11, from ``typing_extensions``) so
>     its name will be defined at runtime. There is no analogous fix for
>     ``reveal_locals``. It simply must be removed from the code before the code
>     is run. (Although, technically, if you really didn't want to remove those
>     calls, you could use ``if not typing.TYPE_CHECKING: reveal_locals = lambda:
>     None`` or similar to define the function to something else at runtime.)

Reviewer 2: ```suggestion
    ``reveal_type`` and ``reveal_locals`` are handled specially by mypy during
    type checking, and don't have to be defined or imported.
    
    However, if you want to run your code,
    you'll have to remove any ``reveal_type`` and ``reveal_locals``
    calls from your program or else Python will give you an error at runtime.
    
    Alternatively, you can import ``reveal_type`` from ``typing_extensions``
    or ``typing`` (on Python 3.11 and newer)
```

   #     b: builtins.str
.. note::

   ``reveal_type`` and ``reveal_locals`` are understood by mypy during typechecking,
   and don't have to be imported as functions. However,
   if you don't import them, then they don't exist at runtime! Therefore,
   you'll have to remove any ``reveal_type`` and ``reveal_locals`` calls from your program

>    or else Python will give you an error at runtime about those names being undefined.

Reviewer 3: The ragged right side makes me unhappy :-(

Author: The test it replaces was exceptionally non-ragged, as well! What a shame.

Author: Anyway, I deliberately broke it up by clause, (like eg https://sembr.org/ ) which leads to exceptional raggedness. But theoretically will make diffs easier to look at when someone edits this in the future.

Author: Hopefully this now sparks joy 😉

   Alternatively, you can import ``reveal_type`` from ``typing_extensions``

>    (or, in more recent versions of python, from ``typing``)

Reviewer 3: Does intersphinx work here? Could we link to `typing.reveal_type`?

Author: I don't know why, but this doesn't seem to work for me. Even though the intersphinices to [sys.version_info](https://docs.python.org/3/library/sys.html#sys.version_info) and typing.TYPE_CHECKING seem to work, typing.reveal_type won't generate for me. Maybe I'm doing something wrong locally and there's some caching problem I'm running into. Anyway, it's harmless to include an intersphinx to typing.reveal_type here because it gracefully degrades down to `normal code`, so I will.

Author: Ah, nevermind: I actually get a hard error when I try to push that: `py:data reference target not found: typing.reveal_type`

Reviewer 3: Unfortunately I can't mark this as resolved :(

Author: I'm not sure if you mean "can't" as in "according to your own principles" or "the UI doesn't let you". I'm assuming the latter, and the UI does let me, so I've done that 🙂.

   so its name will be defined at runtime.
   There is no analogous fix for ``reveal_locals``.
   It simply must be removed from the code before the code is run.
   (Although, technically, if you really didn't want to remove those calls, you could use
   ``if not typing.TYPE_CHECKING: reveal_locals = lambda: None``

>    or similar to define the function to something else at runtime.)

Reviewer 3: IMO this is unnecessary detail and people could come up with this if they need it.

Author: I thought about that when I was writing it, but I like it when documentation errs on the side of telling me too much, instead of too little (which is the dominant failure mode of documentation). And, besides: anyone reading "Common issues and solutions"§"Displaying the type of an expression" probably wants all the help they can get.

Reviewer 3: Makes sense, unfortunately I can't mark this as resolved.
